### PR TITLE
Use pydantic for viewer model

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR follows on from #2042 by starting to use pydantic for the viewermodel. It also includes #2003, #2037, #2040 as they are all working towards making this easier.

It's largely gone well. It's still a bit rough around the edges, and there are a couple things still to do. Probably the largest thing which I haven't been able to get working so nicely is how to handle the `MousemapProvider` and `KeymapProvider` as mixins. Some things that I would have hoped worked didn't quite, but I do think it will be doable.

We'll definitely want to merge those other PRs first, but am putting this here mainly to demonstrate that moving over the viewermodel will be possible too

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
Note this PR removes the event deprecation machinery, which we're not actively using, because of the following error

```python-traceback
(base) czirwc1macos2701:napari nsofroniew$ python examples/add_image.py
Traceback (most recent call last):
  File "examples/add_image.py", line 6, in <module>
    import napari
  File "/Users/nsofroniew/GitHub/napari/napari/__init__.py", line 7, in <module>
    from .viewer import Viewer  # isort:skip
  File "/Users/nsofroniew/GitHub/napari/napari/viewer.py", line 1, in <module>
    from .components import ViewerModel
  File "/Users/nsofroniew/GitHub/napari/napari/components/__init__.py", line 24, in <module>
    from . import _viewer_key_bindings  # isort:skip
  File "/Users/nsofroniew/GitHub/napari/napari/components/_viewer_key_bindings.py", line 3, in <module>
    from .viewer_model import ViewerModel
  File "/Users/nsofroniew/GitHub/napari/napari/components/viewer_model.py", line 47, in <module>
    class ViewerModel(KeymapProvider, MousemapProvider, EventedModel):
  File "pydantic/main.py", line 262, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 315, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 284, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 350, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 379, in pydantic.fields.ModelField._set_default_and_type
  File "pydantic/fields.py", line 287, in pydantic.fields.ModelField.get_default
  File "pydantic/utils.py", line 624, in pydantic.utils.smart_deepcopy
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/copy.py", line 281, in _reconstruct
    if hasattr(y, '__setstate__'):
  File "/Users/nsofroniew/GitHub/napari/napari/utils/events/event.py", line 690, in __getattr__
    if name in self._deprecated:
  File "/Users/nsofroniew/GitHub/napari/napari/utils/events/event.py", line 690, in __getattr__
    if name in self._deprecated:
  File "/Users/nsofroniew/GitHub/napari/napari/utils/events/event.py", line 690, in __getattr__
    if name in self._deprecated:
  [Previous line repeated 466 more times]
RecursionError: maximum recursion depth exceeded
```

Note that the event system already contains a [WarningEmitter](https://github.com/napari/napari/blob/79ab3589eb398ce56d2d0a16025cdb7f435ffaa5/napari/utils/events/event.py#L603-L632) which we're also not using yet. We'll probably want to think about how to deprecate properties and events from our new eventedmodel, but will likely use machinery like this rather than the current event deprecation method

**Comment by sofroniewn** [comment]
The PR also contains some WIP stuff around the keybindings and mousefunctions that I'm not using right now. I'll need to look into this more tomorrow and fix the remaining failing tests. It should be fairly close to being ready for review, but isn't quite ready yet

**Comment by sofroniewn** [comment]
Ok, this is ready for review now @tlambert03 @jni. I've subclassed and overwritten the json serialization to exclude the layer list and the active layer, it just seemed easiest to exclude them at this point. The active layer we'll want to drop, and the layers will become serializable later.

I've also manually excluded from the dict and json methods all the mouse binding and keybinding attributes. @tlambert03 you might want to look this over in case anything unexpected is happening with instances and classes.

I know we'll want to improve this stuff as we go. But for me this takes a little pressure off the selection model improvements, which can then come after we get layer groups further along and the mouse bindings/ key binding improvements which can then come on their own time.

This PR should be backwards compatible and non-breaking.

**Review by tlambert03 — COMMENTED** [review]
looking pretty good.  still more I want to look at, but in playing with it a bit, I have an updated suggestion for handling the `__init__` sequence.  There is currently another problem with the mixins, because their `__init__` functions themselves call `super().__init__()`, meaning `ViewerModel.__init__` basically gets called again with no arguments.  So probably a better solution here is to fix the mixins themselves to be "better citizens" for subclassing.  So, both `Keymap` and `MousemapProvider` should have this in their init:
```python
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
```

Then, in ViewerModel, we can go back to the standard init style with a single `super()`.  Furthermore, you can go ahead and pass through the arguments that are destined for the `Dims()` model as a dict and remove that whole `# set initial values` part.  You can also just temporarily allow `extra` (and remove the `class Config` override).  So, all told, it looks like this:

```python
class ViewerModel(KeymapProvider, MousemapProvider, EventedModel):
    ...
    def __init__(self, title='napari', ndisplay=2, order=(), axis_labels=()):
        self.__config__.extra = 'allow'
        super().__init__(
            title=title,
            dims={
                'axis_labels': axis_labels,
                'ndisplay': ndisplay,
                'order': order,
            },
        )
        self.__config__.extra = 'ignore'
		...
```

File: napari/components/viewer_model.py [hunk 1]
```diff
@@ -68,41 +81,50 @@ class ViewerModel(KeymapProvider, MousemapProvider):
...
+    active_layer: Optional[
+        Layer
+    ] = None  # Would be nice to remove this/ make it layer name instead of layer?
+    help: str = ''
+    status: str = 'Ready'
+    theme: str = DEFAULT_THEME
+    title: str = 'napari'
+
+    # 2-tuple indicating height and width
+    _canvas_size = (600, 800)
```
Comment (tlambert03) [inline napari/components/viewer_model.py]:
```suggestion
    _canvas_size: Tuple[int, int] = (600, 800)
```

(and add typing import)

File: napari/components/viewer_model.py [hunk 2]
```diff
@@ -68,41 +81,50 @@ class ViewerModel(KeymapProvider, MousemapProvider):
...
+        default_factory=LayerList
+    )  # Need to create custom JSON encoder for layer!
+    scale_bar: ScaleBar = ScaleBar()
+
+    active_layer: Optional[
+        Layer
+    ] = None  # Would be nice to remove this/ make it layer name instead of layer?
+    help: str = ''
+    status: str = 'Ready'
+    theme: str = DEFAULT_THEME
```
Comment (tlambert03) [inline napari/components/viewer_model.py:102]: perhaps we should make a `Theme` enum that gets dynamically populated?
```python
Themes = StringEnum('Themes', [(x.upper(), x) for x in available_themes()])
Themes.DEFAULT = Themes.DARK
```

I believe you could then remove the theme validator below

File: napari/components/viewer_model.py [hunk 3]
```diff
@@ -116,81 +138,44 @@ def __init__(self, title='napari', ndisplay=2, order=(), axis_labels=()):
...
-
-    @property
-    def title(self):
-        """string: String that is displayed in window title."""
-        return self._title
+        return v
+
+    def json(self, **kwargs):
+        """Serialize to json."""
+        # Manually exclude the layer list and active layer which cannot be serialized at this point
```
Comment (tlambert03) [inline napari/components/viewer_model.py:152]: 👍

File: napari/utils/events/event.py [hunk 4]
```diff
@@ -729,12 +724,6 @@ def __init__(
 
     # mypy fix for dynamic attribute access
     def __getattr__(self, name: str) -> EventEmitter:
```
Comment (tlambert03) [inline napari/utils/events/event.py]: if we're removing the warning here... I think you can just remove the whole `__getattr__` method

File: napari/components/viewer_model.py (hunk 2 above)
Comment (sofroniewn) [reply napari/components/viewer_model.py:102]: What if someone adds a theme during run time?

File: napari/components/viewer_model.py [hunk 5]
```diff
@@ -68,41 +82,47 @@ class ViewerModel(KeymapProvider, MousemapProvider):
...
+    ] = None  # Would be nice to remove this/ make it layer name instead of layer?
+    help: str = ''
+    status: str = 'Ready'
+    theme: str = DEFAULT_THEME
+    title: str = 'napari'
+
+    # 2-tuple indicating height and width
+    _canvas_size: Tuple[int, int] = (600, 800)
+
+    class Config(EventedModel.Config):
```
Comment (sofroniewn) [inline napari/components/viewer_model.py]: I wasn't able to use the `self.__config__.extra = 'allow'` syntax in the init, it didn't seem to work for me. Not sure if I was getting something trivial wrong, but I tried with what you posted in comment above

File: napari/components/viewer_model.py [hunk 6]
```diff
@@ -68,41 +82,47 @@ class ViewerModel(KeymapProvider, MousemapProvider):
...
+        # allow extra attributes during model initialization, useful for mixins
+        extra = 'allow'
 
-        self.dims = Dims(
-            ndisplay=ndisplay, order=order, axis_labels=axis_labels
+    def __init__(self, title='napari', ndisplay=2, order=(), axis_labels=()):
+        super().__init__(
+            title=title,
```
Comment (sofroniewn) [inline napari/components/viewer_model.py:112]: love this now, much nicer

File: napari/utils/events/event.py (hunk 4 above)
Comment (sofroniewn) [reply napari/utils/events/event.py]: dropped

**Review by sofroniewn — COMMENTED** [review]
@tlambert03 I've responded to your initial round of review

File: napari/components/viewer_model.py (hunk 2 above)
Comment (tlambert03) [reply napari/components/viewer_model.py:102]: good point.  ok, let's leave that for later.  Theme validation shouldn't be on the viewer model though...  but we can make a proper theme model later.

File: napari/components/viewer_model.py (hunk 5 above)
Comment (tlambert03) [reply napari/components/viewer_model.py]: huh... what error did you get?  You can also try `self.__config__.extra = Extra.allow` with `from pydantic import Extra`

File: napari/components/viewer_model.py (hunk 2 above)
Comment (sofroniewn) [reply napari/components/viewer_model.py:102]: yeah sounds good., i feel like with settings coming there could more improvements to how we handle this in the future, but it's fine for now

File: napari/components/viewer_model.py (hunk 5 above)
Comment (sofroniewn) [reply napari/components/viewer_model.py]: Yup, that's it. Using just the string the extra attributes just don't get added, no special error message until you try and use one. Using the Enum works. I can use that syntax now

**Review by tlambert03 — APPROVED** [review]
i think i'm happy with it!

**Comment by sofroniewn** [comment]
> i think i'm happy with it!

Nice, I will review this tomorrow morning, unless i hear otherwise from @jni!!

**Pull request merged**